onmt/utils/Cuda.lua

```diff
@@ -52,16 +52,21 @@ end
   When using CPU only, converts to float instead of the default double.
 ]]
 function Cuda.convert(obj)
-  if torch.typename(obj) then
+  local objtype = torch.typename(obj)
+  if objtype then
     if Cuda.activated and obj.cuda ~= nil then
-      return obj:cuda()
-    elseif not Cuda.activated and obj.float ~= nil then
+      if objtype:find('torch%..*LongTensor') then
+        return obj:cudaLong()
+      else
+        return obj:cuda()
+      end
+    elseif not Cuda.activated and obj.float ~= nil and not objtype:find('torch%..*LongTensor') then
```

Comment:
Close. :)

`CudaLongTensor` are not moved back to CPU with this test. The piece of code I shared handled this case.